test-data/unit/check-overloading.test

```diff
@@ -1136,3 +1136,24 @@ def f(a: Callable[[], int]) -> None: pass
 @overload
 def f(a: str) -> None: pass
 f(0)  # E: No overload variant of "f" matches argument types [builtins.int]
+
+[case testCustomRedefinitionDecorator]
+from typing import Any, Callable, Type
+
+class Chain(object):
+    def chain(self, function: Callable[[Any], int]) -> 'Chain':
+        return self
+
+class Test(object):
+    do_chain = Chain()
+
+    @do_chain.chain
+    def do_chain(self) -> int:
+        return 2
+
+    @do_chain.chain  # E: Name 'do_chain' already defined
```

Comment:
(Response is mostly to Guido, not Ivan)

Ok, I've now read enough code to understand why we get redefinition errors when we do: we suppress them when checking overloads, and then re-produce them if it turns out what we were checking wasn't a "real" overload. 

If this kind of chaining behaviour (or other situations where a decorator makes redefinition of a function the desired idiom) is not a fluke and happens with any frequency in the wild, I have a particular code block to delete and some tests to change. This is the option I'm leaning towards, and can have a PR that does it soon. 

If we keep wanting to say you need one of our special decorators (or a special name, cf Ivan's convent) to make redefinition ok, I can fix the bug that suppresses the first of those errors when the initial definition is by assignment. It's a minor bug, in the scheme of things. 

Either way, I think you should accept this PR as-is, fixing the cash for now.